src/internet_identity/src/main.rs

```diff
@@ -363,36 +374,43 @@ fn update_root_hash() {
 
 /// Authenticates the caller (traps if not authenticated) and updates the device used to authenticate
 /// reflecting the current activity. Also updates the aggregated stats on daily and monthly active users.
+///
+/// Note: this function reads / writes the anchor from / to stable memory. It is intended to be used by functions that
+/// do not further modify the anchor.
 fn authenticate_and_record_activity(anchor_number: AnchorNumber) {
-    let anchor = state::anchor(anchor_number);
+    let mut anchor = state::anchor(anchor_number);
     let device_key = trap_if_not_authenticated(&anchor).pubkey.clone();
-    let previous_activity = anchor.last_activity();
-    anchor_management::update_last_device_usage(anchor_number, anchor, &device_key);
-    active_anchor_stats::update_active_anchors_stats(previous_activity);
+    anchor_management::activity_bookkeeping(&mut anchor, &device_key);
+    state::storage_mut(|storage| storage.write(anchor_number, anchor)).unwrap_or_else(|err| {
+        panic!("last_usage_timestamp update: unable to update anchor {anchor_number}: {err}")
+    })
 }
 
 /// Authenticates the caller (traps if not authenticated) calls the provided function and handles all
 /// the necessary bookkeeping for anchor operations.
-fn authenticated_anchor_operation(
+fn authenticated_anchor_operation<R>(
```

Comment:
Also here documenting the intended role of `R` and the meaning of the result of `op` would be useful.

Reply:
Maybe extend the comment, with an explanation that `R` is usually bound to a candid-type, which contains both successful and failure variants, hence it can be used in both positions of `Result<T, E>`?